manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,790 @@
+
+The OCaml memory model guarantees that even for programs with data races,
+memory safety is preserved. While programs with data races may observe
+non-sequentially consistent behaviours, they will not crash. Let us look at an
+example of how to utilise LDRF to argue about memory safety in the presence of
+data races. Let us take a variant of the message passing example seen earlier
+(section \ref{s:mm_msg_passing}).
+
+\begin{caml_example*}{verbatim}
+let r = ref None
+let d1 () =
+  r := Some "Hello, world!"
+let d2 () =
+  match !r with
+  | None -> ()
+  | Some s -> print_endline s
+\end{caml_example*}
+
+The program uses a non-atomic reference "r" to share a message between domains.
+It is easy to see that there is a data race on "r". For the value "Some
+\"Hello, world!\"", OCaml allocates two objects -- one for the string and the
+other for the "Some" constructor. Both of these objects are first allocated
+uninitialised and then initialised with initialising writes. Are we guaranteed
+that, when "d2" sees the "Some" variant when reading "r", the objects have been
+fully initialised?
+
+The answer is yes, and we can use LDRF to reason about this. Let us refactor the
+program to make the newly allocated objects explicit.
+
+\begin{caml_example*}{verbatim}
+let r = ref None
+let d1 () =
+  let s = "Hello, world!" in
+  let o = Some s in
+  r := o
+let d2 () =
+  let o = !r in
+  match o with
+  | None -> ()
+  | Some s -> print_endline s
+\end{caml_example*}
+
+Observe that if "d2" does not witness the initialising writes on "s" or "o",
+the program may crash. LDRF allows us to reason about the program behaviour for
+a subset of memory locations. Let us ignore "r" and only consider "s" and "o".
+The objects "s" and "o" have initialising writes in "d1" and reads in "d2". The
+happens-before definition (section \ref{s:happens_before}) says that the
+initial write to a location precedes any other actions on that location. Hence,
+there is a happens-before relationship between the initial writes to "s" and
+"o" in "d1" and the reads in "d2". Hence, there is no data race on "s" and "o",
+and the program fragment that uses "s" and "o" will have sequentially
+consistent behaviour.
+
+Since "s" and "o" are newly allocated in "d1", the only schedule where "d2"
+will read those objects is the one in which "d1" executes before "d2". Thanks
+to LDRF-SC, we can use sequential reasoning, which guarantees that "d2" will
+always see the initialising writes from "d1".
+
+\section{s:mm_semantics}{An operational view of the memory model}
+
+In this section, we describe the semantics of the OCaml memory model. Given an
+OCaml program, which may possibly contain data races, the semantics tells you
+the values that may be observed by the read of a memory location. For
+simplicity, we restrict the intra-thread actions to just the accesses to atomic
+and non-atomic locations, ignoring domain spawn and join operations, and the
+operations on mutexes.
```

Comment:
Why so? Does it really complicate the model so much?

Reply:
Yes, I would think so. The operational semantics

Reply:
I don't have a particularly strong opinion on this and I don't want to block anything, but I indeed think that the PLDI 2018 presentation is lacking some important aspects. I think it would be a good think to include such a more in-depth description here because:
- As experience shows (e.g., allocations), the memory model for the full language is never a "trivial extension", and subtle problems can arise unexpectedly (e.g., how one should formulate the model to make sure that a read is always safe even though it is not happens-after the allocation of the cell...). I understand that this text is targeted at OCaml developpers which are not necessarily interested in all these details, but if it's not described in the OCaml manual, then where else should it be described?
- I don't think adding explanations for spawn/join/allocations/mutexes/... is invasive in the sense that it would require changing the text for the other operations. The general setting (frontiers...) need not be changed.

Reply:
> As experience shows (e.g., allocations), the memory model for the full language is never a "trivial extension"

Indeed. This matched our experience of implementing the model. I have a sufficient understanding of how allocation and initialisation work in the memory model now. 

Perhaps the right approach would be to first write a peer-reviewed paper that covers allocations and initialisations, compilation to LKMM and reasoning about concurrent accesses to shared locations from OCaml (hardware model) and C (LKMM). And then we can aim to summarize that in the manual. I am hesitant to include new tricky material on the memory model directly in the manual. What do you think?

> I don't think adding explanations for spawn/join/allocations/mutexes/... is invasive in the sense that it would require changing the text for the other operations. The general setting (frontiers...) need not be changed.

In a follow-up PR, I'll try to extend the operational semantics with spawn, join, mutexes, condition and semaphore.

Reply:
> Perhaps the right approach would be to first write a peer-reviewed paper that covers allocations and initialisations, compilation to LKMM and reasoning about concurrent accesses to shared locations from OCaml (hardware model) and C (LKMM). And then we can aim to summarize that in the manual. I am hesitant to include new tricky material on the memory model directly in the manual. What do you think?

Seems good to me. The memory model of OCaml and its interaction with C are two different matters.

> Could you point me to memory model papers that explicitly deal with allocations?

Our work on Cosmo deals with allocations.

> In a follow-up PR, I'll try to extend the operational semantics with spawn, join, mutexes, condition and semaphore.

:+1: